### PR TITLE
Move `sample_mask` to `transform` method in `BaseMasker`, handle `sample_mask` in `signal.clean`

### LINKED ISSUE #1012: Implement sample mask in signal cleaning and the various maskers
The goal of this issue is partly to implement/support scrubbing of time series.

This issue depends on #1011

### PR DESCRIPTION
**Pull request by htwangtw**
Closes #1012.

Changes proposed in this pull request:

- In `base_masker`, move parameter `sample_mask` from initialization to `transform` method, and all the Masker modules.
- Handle `sample_masks` with `signal.clean`
- Move input validation related to `runs` to a separate private function from `_process_runs`  and `_sanitize_inputs` to `_sanitize_runs` 

While implementing the discussed changes, I found how sample_mask should be handled when there are multiple runs more complicated than anticipated.
Here's a summary of the current implementation:  
* accept any input valid for indexing NumPy array as `sample_mask`
* for multiple runs, user must pass a list of sample masks to `sample_mask` (a set of index per run).

I have tried to implement `sample_mask` to match `confounds` for multiple runs, which allows sample_mask to be a 1D array when runs are provided. User needs to correct the index for multiple runs to fit into a 1D array. I think it's too error-prone. This makes me wonder, should we do the same for confounds? Or we want to rethink how multiple runs are handled?

This PR has a lot of difficult things to handle. If there's any redundant checks in signal.clean, or any better idea of handling the `sample_mask`, `runs`, and `confounds` please let me know.

### CODE REVIEW

File: nilearn/signal.py [hunk 1]
```diff
@@ -71,7 +72,9 @@ def _standardize(signals, detrend=False, standardize='zscore'):
                 signals = signals - signals.mean(axis=0)
 
             std = signals.std(axis=0)
-            std[std < np.finfo(np.float64).eps] = 1.  # avoid numerical problems
+            std[
+                std < np.finfo(np.float64).eps
+            ] = 1.0  # avoid numerical problems
```
Comment (bthirion) [inline nilearn/signal.py:77]: I find this less readable, but maybe it's a matter of taste. Was the line too long ?

**Review by bthirion — COMMENTED** [review]
LGTM otherwise. Great work !

**Comment by bthirion** [comment]
> I have tried to implement `sample_mask` to match `confounds` for multiple runs, which allows sample_mask to be a 1D array when runs are provided. User needs to correct the index for multiple runs to fit into a 1D array. I think it's too error-prone. This makes me wonder, should we do the same for confounds? Or we want to rethink how multiple runs are handled?

IMHO the story is a bit different here for `confounds` and `sample_mask`. Indeed, in `sample_mask` there is the additional indexing difficulty.
I think it is acceptable not to handle 1D array for `sample mask`, given the ambiguities it entails, but I would keep it for `confounds` for the sake of bw compatibility.
Thx !

File: nilearn/input_data/multi_nifti_masker.py [hunk 2]
```diff
@@ -185,49 +199,60 @@ def fit(self, imgs=None, y=None):
...
+                compute_mask, ignore=['n_jobs', 'verbose', 'memory']
+            )(
+                imgs,
+                target_affine=self.target_affine,
+                target_shape=self.target_shape,
+                n_jobs=self.n_jobs,
+                memory=self.memory,
+                verbose=max(0, self.verbose - 1),
+                **mask_args
+            )
```
[the remark this answers is not in the document]
Comment (htwangtw) [reply nilearn/input_data/multi_nifti_masker.py:247]: I was trying to format the code as flake8 was blocking the test in CI. I did realise that was a bad idea.

**Comment by htwangtw** [comment]
Close this PR and replace it with a new one that's less messed up by automatic linting

**Comment by bthirion** [comment]
Sorry for the additional work !

**Comment by htwangtw** [comment]
Not at all!

**Pull request closed without merging**